2.9BSD/usr/man/cat8/init.8

INIT(8)             UNIX Programmer's Manual              INIT(8)

NAME
     init - process control initialization

SYNOPSIS
     /etc/init

DESCRIPTION
     _I_n_i_t is invoked inside UNIX as the last step in the boot
     procedure.  If this is a boot from software (automatic or
     from _r_e_b_o_o_t(8), it normally then runs the automatic reboot
     sequence as described in _r_e_b_o_o_t(8)), and if this succeeds,
     begins multi-user operation.  If the reboot fails, it com-
     mences ``special session'' operation on a limited number of
     terminals (see below).  It is possible to pass parameters
     from the boot program to _i_n_i_t so that single user operation
     is commenced immediately.  When such single user operation
     is terminated by killing the single-user shell (i.e. by hit-
     ting ^D), _i_n_i_t runs /_e_t_c/_r_c without the reboot parameter.
     This command file performs housekeeping operations such as
     removing temporary files, mounting file systems, and start-
     ing daemons.

     In multi-user operation, _i_n_i_t'_s role is to create a process
     for each terminal port on which a user may log in.  To begin
     such operations, it reads the file /_e_t_c/_t_t_y_s and forks
     several times to create a process for each terminal speci-
     fied in the file.  Each of these processes opens the
     appropriate terminal for reading and writing.  These chan-
     nels thus receive file descriptors 0, 1 and 2, the standard
     input and output and the diagnostic output.  Opening the
     terminal will usually involve a delay, since the _o_p_e_n is not
     completed until someone is dialed up and carrier established
     on the channel.  If a terminal exists but an error occurs
     when trying to open the terminal _i_n_i_t complains by writing a
     message to the system console; the message is repeated every
     10 minutes for each such terminal until the terminal is shut
     off in /_e_t_c/_t_t_y_s and init notified (by a hangup, as
     described below), or the terminal becomes accessible (init
     checks again every minute).  After an open succeeds,
     /_e_t_c/_g_e_t_t_y is called with argument as specified by the
     second character of the _t_t_y_s file line.  _G_e_t_t_y reads the
     user's name and invokes _l_o_g_i_n to log in the user and execute
     the Shell.

     Ultimately the Shell will terminate because of an end-of-
     file either typed explicitly or generated as a result of
     hanging up.  The main path of _i_n_i_t, which has been waiting
     for such an event, wakes up and removes the appropriate
     entry from the file _u_t_m_p, which records current users, and
     makes an entry in /_u_s_r/_a_d_m/_w_t_m_p, which maintains a history
     of logins and logouts.  The _w_t_m_p entry is made only if a
     user logged in successfully on the line.  Then the

Printed 8/2/83                                                  1

INIT(8)             UNIX Programmer's Manual              INIT(8)

     appropriate terminal is reopened and _g_e_t_t_y is reinvoked.

     _I_n_i_t catches the _h_a_n_g_u_p signal (signal SIGHUP) and inter-
     prets it to mean that the file /_e_t_c/_t_t_y_s should be read
     again.  The Shell process on each line which used to be
     active in _t_t_y_s but is no longer there is terminated; a new
     process is created for each added line; lines unchanged in
     the file are undisturbed.  Thus it is possible to drop or
     add phone lines without rebooting the system by changing the
     _t_t_y_s file and sending a _h_a_n_g_u_p signal to the _i_n_i_t process:
     use `kill -HUP 1.'

     _I_n_i_t will terminate multi-user operations and resume
     single-user mode if sent a terminate (TERM) signal, i.e.
     ``kill -TERM 1''.  If there are processes outstanding which
     are deadlocked (due to hardware or software failure), _i_n_i_t
     will not wait for them all to die (which might take for-
     ever), but will time out after 30 seconds and print a warn-
     ing message.

     _I_n_i_t will cease creating new _g_e_t_t_y's and allow the system to
     slowly die away, if it is sent a quit signal (QUIT) signal,
     i.e. ``kill -QUIT 1''.  A later hangup will resume full
     multi-user operations, or a terminate will initiate a single
     user shell.  This hook is used by _r_e_b_o_o_t(8).

     ``Special session'' is similar to multi-user operation,
     except that only root may log in, and a only a subset of the
     normal lines are used (those with `2' or `3' in the first
     position of their entries in /_e_t_c/_t_t_y_s).  _I_n_i_t enters this
     mode if an automatic file system check fails at boot time,
     allowing the necessary file system repairs to be made.  When
     _i_n_i_t then receives an interrupt signal (INT), it shuts down,
     runs /_e_t_c/_r_c, then comes up for normal multi-user opera-
     tions.  An interrupt signal from multi-user mode shuts down
     the system, then begins special session.

     _I_n_i_t'_s role is so critical that if it dies, the system will
     reboot itself automatically.  If, at bootstrap time, the
     _i_n_i_t process cannot be located, the system will print a mes-
     sage and loop in user mode at location 06.

DIAGNOSTICS
     Diagnostic messages from _i_n_i_t are printed on the system con-
     sole.

     init: _t_t_y: cannot open.  A terminal which is turned on in
     the _t_t_y_s file cannot be opened, likely because the requisite
     lines are either not configured into the system or the asso-
     ciated special files in /dev do not exist.

Printed 8/2/83                                                  2

INIT(8)             UNIX Programmer's Manual              INIT(8)

     WARNING: Something is hung (won't die); ps axl advised.  A
     process is hung and could not be killed when the system was
     shutting down.  This is usually caused by a process which is
     stuck in a device driver due to a persistent device error
     condition.

FILES
     /dev/console        for diagnostic messages
     /dev/tty??
     /etc/utmp
     /usr/adm/wtmp
     /etc/ttys
     /etc/rc             script run when going multiuser

SEE ALSO
     kill(1), login(1), sh(1), ttys(5), crash(8), getty(8),
     rc(8), reboot(8), shutdown(8)

Printed 8/2/83                                                  3



2.9BSD/usr/src/cmd/init.c
/*
 *	init.c
 */
#include "whoami.h"
#include <signal.h>
#include <sys/types.h>
#include <utmp.h>
#include <setjmp.h>
#ifdef	UCB_AUTOBOOT
#include <sys/reboot.h>
#endif
#include <errno.h>
#include <sys/autoconfig.h>

#define	VHANGUP			/* undefine if you don't have vhangup */
#define	CONFIGOPTS	"-vc"

#define	LINSIZ	sizeof(wtmp.ut_line)
#define	TABSIZ	100
#define	ALL	p = &itab[0]; p < &itab[TABSIZ]; p++
#define	EVER	;;
#define SCPYN(a, b)	strncpy(a, b, sizeof(a))
#define SCMPN(a, b)	strncmp(a, b, sizeof(a))

char	shell[]	= "/bin/sh";
char	getty[]	 = "/etc/getty";
char	minus[]	= "-";
char	plus[] = "+";		/* this tells getty we are in sp. sess. */
char	runc[]	= "/etc/rc";
char	ifile[]	= "/etc/ttys";
char	utmp[]	= "/etc/utmp";
char	wtmpf[]	= "/usr/adm/wtmp";
char	ctty[]	= "/dev/console";
char	dev[]	= "/dev/";
char	config[]= "/etc/autoconfig";

struct utmp wtmp;
struct
{
	char	line[LINSIZ];
	char	comn;
	char	flag;
} line;
struct	tab
{
	char	line[LINSIZ];
	char	comn;
	char	xflag;
	int	pid;
} itab[TABSIZ];

extern	errno;

int	fi;
int	mergflag;
int	multiuser;
/*
 *  Modes: bits in first char of /etc/ttys line
 */
#define	NORMAL	1		/* anyone can login */
#define	SP_SESS	2		/* only root can login */
int	mode = NORMAL;

char	tty[20];
jmp_buf	sjbuf, shutpass;
time_t	time0;

time_t	time();
int	reset();
int	idle(), sp_ss(), setmerge();
char	*strcpy(), *strcat();
long	lseek();

#ifdef	UCB_AUTOBOOT
main(ac, av)
char	**av;
#else
main ()
#endif
{
	int howto, oldhowto;

	time0 = time((time_t *) 0);
#ifdef	UCB_AUTOBOOT
	signal(SIGQUIT, idle);

	if (ac > 1) {
		howto = * ((int *) av[1]);
		* ((int *) av[1]) = 0;	/* don't confuse ps with binary args */
	 } else
		howto = RB_SINGLE;
	if (autoconfig() == 0)
		howto = RB_SINGLE;
#else
	autoconfig();
#endif
	setjmp(sjbuf);
	signal(SIGTERM, reset);
	signal(SIGINT, sp_ss);
	signal(SIGHUP, setmerge);
	for(EVER) {
		shutdown();
#ifdef	UCB_AUTOBOOT
		oldhowto = howto;
		howto = RB_SINGLE;
		if (oldhowto & RB_SINGLE)
			single();
		if (runcom(oldhowto) == 0) 
			mode = SP_SESS;
#else
		single();
		runcom();
#endif
		merge();
		multiple();
	}
}

int	shutreset();

shutdown()
{
	register i, f;
	register struct tab *p;

	multiuser = 0;
	for(ALL) {
		term(p);
		p->line[0] = 0;
	}
	close(creat(utmp, 0644));
	signal(SIGALRM, shutreset);
	if (setjmp(shutpass) == 0) {
		alarm(30);
		for(i=0; i<5; i++)
			kill(-1, SIGKILL);
		while(wait((int *)0) != -1)
			;
		alarm(0);
	}
	acct(0);
	signal(SIGALRM, SIG_DFL);
	for(i=0; i<10; i++)
		close(i);
	f = open(wtmpf, 1);
	if (f >= 0) {
		lseek(f, 0L, 2);
		SCPYN(wtmp.ut_line, "~");
		SCPYN(wtmp.ut_name, "shutdown");
		time(&wtmp.ut_time);
		write(f, (char *)&wtmp, sizeof(wtmp));
		close(f);
	}
}

shutreset()
{
	cmesg("WARNING: Something is hung (won't die); ps axl advised\n", 0, 0);
	longjmp(shutpass, 1);
}

single()
{
	register pid;
	register xpid;
	extern	errno;

	multiuser = 0;
   do {
	pid = fork();
	if(pid == 0) {
/*
		alarm(300);
*/
		signal(SIGTERM, SIG_DFL);
		signal(SIGHUP, SIG_DFL);
		signal(SIGALRM, SIG_DFL);
		open(ctty, 2);
		dup(0);
		dup(0);
		execl(shell, minus, (char *)0);
		cmesg("Init: can't exec ", shell, "\r\n");
		exit(0);
	}
	while((xpid = wait((int *)0)) != pid)
		if (xpid == -1 && errno == ECHILD)
			break;
   } while (xpid == -1);
}

#ifdef	UCB_AUTOBOOT
runcom(howto)
int howto;
#else
runcom()
#endif
{
	register pid, f;
	int status;
	char *arg1, *arg2;

	pid = fork();
	if(pid == 0) {
		open("/", 0);
		dup(0);
		dup(0);
#ifdef	UCB_AUTOBOOT
		if ((howto & RB_SINGLE) || (howto & RB_NOFSCK))
			arg1 = "fastboot";
		else
			arg1 = "autoboot";
		if (howto & RB_POWRFAIL)
			arg2 = "powerfail";
		else
			arg2 = (char *)0;
		execl(shell, shell, runc, arg1, arg2, (char *)0);
		exit(1);
#else
		execl(shell, shell, runc, (char *)0);
		exit(1);
#endif
	}
	while(wait(&status) != pid)
		;
#ifdef	UCB_AUTOBOOT
	if(status)
		return(0);
#endif
	f = open(wtmpf, 1);
	if (f >= 0) {
		lseek(f, 0L, 2);
		SCPYN(wtmp.ut_line, "~");
		SCPYN(wtmp.ut_name, "reboot");
		if (time0) {
			wtmp.ut_time = time0;
			time0 = 0;
		} else
			time(&wtmp.ut_time);
		write(f, (char *)&wtmp, sizeof(wtmp));
		close(f);
	}
	return(1);
}

setmerge()
{
	signal(SIGHUP, SIG_IGN);
	mergflag = 1;
}

multiple()
{
	register struct tab *p;
	register pid;

loop:
	multiuser = 1;
	mergflag = 0;
	signal(SIGHUP, setmerge);
	for(EVER) {
		pid = wait((int *)0);
		if(mergflag) {
			merge();
			goto loop;
		}
		if(pid == -1) {
			if (errno == ECHILD) {
				cmesg("Init: ", "no children left", "\r\n");
				return;
			}
			goto loop;
		}
		for(ALL)
			if(p->pid == pid || p->pid == -1) {
#ifdef	UCB_SUBMIT
				if (p->pid != -1)
					killbkg(p->pid, SIGKILL);
#endif
				rmut(p);
				dfork(p);
			}
	}
}

term(p)
register struct tab *p;
{

	if(p->pid != 0 && p->pid != -1) {
		rmut(p);
		kill(p->pid, SIGKILL);
#ifdef	UCB_SUBMIT
		killbkg(p->pid, SIGKILL);
#endif
	}
	p->pid = 0;
}

rline()
{
	register c, i;

loop:
	c = get();
	if(c < 0)
		return(0);
	if(c == 0)
		goto loop;
	line.flag = c;
	c = get();
	if(c <= 0)
		goto loop;
	line.comn = c;
	SCPYN(line.line, "");
	for (i=0; i<LINSIZ; i++) {
		c = get();
		/*
		 * If a blank, newline, or end of file, or tab,
		 * cease accumulating line name
		 */
		if(c <= 0 || c == ' ' || c == '\t')
			break;
		line.line[i] = c;
	}
	while(c > 0)
		c = get();
	if(line.line[0] == 0)
		goto loop;
	if(line.flag == '0')
		goto loop;
	strcpy(tty, dev);
	strncat(tty, line.line, LINSIZ);
	if(access(tty, 06) < 0)
		goto loop;
	return(1);
}

get()
{
	char b;

	if(read(fi, &b, 1) != 1)
		return(-1);
	if(b == '\n')
		return(0);
	return(b);
}

#define	FOUND	1
#define	CHANGE	2

merge()
{
	register struct tab *p;

	fi = open(ifile, 0);
	if(fi < 0)
		return;
	for(ALL)
		p->xflag = 0;
	while(rline()) {
		if ((line.flag < '1') || (line.flag > '9'))
			continue;
		if (((line.flag-'0') & mode) == 0)
			continue;
		for(ALL) {
			if (SCMPN(p->line, line.line))
				continue;
			p->xflag |= FOUND;
			if(line.comn != p->comn) {
				p->xflag |= CHANGE;
				p->comn = line.comn;
			}
			goto contin1;
		}
		for(ALL) {
			if(p->line[0] != 0)
				continue;
			SCPYN(p->line, line.line);
			p->xflag |= FOUND|CHANGE;
			p->comn = line.comn;
			goto contin1;
		}
	contin1:
		;
	}
	close(fi);
	for(ALL) {
		if((p->xflag&FOUND) == 0) {
			term(p);
			p->line[0] = 0;
		}
		if((p->xflag&CHANGE) != 0) {
			term(p);
			dfork(p);
		}
#ifdef	UCB_AUTOBOOT
		/*
		 * If we are resuming after an idle (possibly a failed reboot)
		 * we need to restart the lines that shut down.
		 */
		if (p->pid == -1)
			dfork(p);
#endif
	}
}

dfork(p)
struct tab *p;
{
	register pid;

	pid = fork();
	if(pid == 0) {
		signal(SIGTERM, SIG_DFL);
		signal(SIGHUP, SIG_IGN);
		strcpy(tty, dev);
		strncat(tty, p->line, LINSIZ);
		chown(tty, 0, 0);
		chmod(tty, 0622);
		if (open(tty, 2) < 0) {
			int repcnt = 0;
			do {
				if (repcnt % 10 == 0)
					cmesg("init: ",tty,": cannot open\n\r");
				repcnt++;
				sleep(60);
			} while (open(tty, 2) < 0);
		}
#ifdef	VHANGUP
		vhangup();
#endif
		signal(SIGHUP, SIG_DFL);
		open(tty, 2);
		close(0);
		dup(1);
		dup(0);
		tty[0] = p->comn;
		tty[1] = 0;
		if(mode == SP_SESS)
			execl(getty, plus, tty, (char *)0);
		else
			execl(getty, minus, tty, (char *)0);
		exit(0);
	}
	p->pid = pid;
}

rmut(p)
register struct tab *p;
{
	register f;
	int found = 0;

	f = open(utmp, 2);
	if(f >= 0) {
		while(read(f, (char *)&wtmp, sizeof(wtmp)) == sizeof(wtmp)) {
			if (SCMPN(wtmp.ut_line, p->line) || wtmp.ut_name[0]==0)
				continue;
			lseek(f, -(long)sizeof(wtmp), 1);
			SCPYN(wtmp.ut_name, "");
			time(&wtmp.ut_time);
			write(f, (char *)&wtmp, sizeof(wtmp));
			found++;
		}
		close(f);
	}
	if (found) {
		f = open(wtmpf, 1);
		if (f >= 0) {
			SCPYN(wtmp.ut_line, p->line);
			SCPYN(wtmp.ut_name, "");
			time(&wtmp.ut_time);
			lseek(f, (long)0, 2);
			write(f, (char *)&wtmp, sizeof(wtmp));
			close(f);
		}
	}
}

reset()
{
	longjmp(sjbuf, 1);
}

/*
 *  Toggle special-session mode.
 *  Do a shutdown() so that getty's are reissued in the new mode.
 */
sp_ss()
{
	signal(SIGINT, SIG_IGN);
	mergflag++;
	shutdown();
	if (mode == NORMAL) 
		mode = SP_SESS;
	else {
		/*
		 * Returning to normal operation.
		 * Run the rc file; either it hasn't been finished
		 * since we failed a filesystem check, or we shut down.
		 */
#ifdef	UCB_AUTOBOOT
		(void) runcom(RB_NOFSCK);
#else
		(void) runcom();
#endif
		mode = NORMAL;
	}
	signal(SIGINT, sp_ss);
}

#ifdef	UCB_AUTOBOOT
idle()
{
	register struct tab *p;
	register pid;

	signal(SIGQUIT,idle);
	for (;;) {
		pid = wait((int *) 0);
		if (mergflag) {
			if (!multiuser)
				reset();
			else
				return;
		}
		if (pid == -1)
			pause();
		else {
			for (ALL)
				if (p->pid == pid) {
					rmut(p);
					p->pid = -1;
				}
		}
	}
}
#endif

cmesg(s1, s2, s3)
char *s1, *s2, *s3;
{
	register int pid;

	pid = fork();
	if (pid == 0) {
		int fd = open(ctty, 2);
		write(fd, s1, strlen(s1));
		if (s2)
			write(fd, s2, strlen(s2));
		if (s3)
			write(fd, s3, strlen(s3));
		close(fd);
		exit(0);
	}
	while (wait((int *)0) != pid)
		;
}

autoconfig()
{
	int pid, status;

	cmesg("\r\nCONFIGURE SYSTEM:\n", 0, 0);
	if ((pid = fork()) == 0) {
		open(ctty, 2);
		dup(0);
		dup(0);
		execl(config, "autoconfig", CONFIGOPTS, 0);
		printf("Couldn't exec %s\n", config);
		exit(AC_SETUP);
	}
	while (wait(&status) != pid)
		;
	if ((status & 0377) == 0)
		status >>= 8;
	else
		status = AC_SINGLE;
	switch (status) {
		case AC_SETUP:
			cmesg("Configuration setup error\n", 0, 0);
			return 0;
		case AC_SINGLE:
			cmesg("SERIOUS CONFIGURATION ERROR\n", 0, 0);
			return 0;
		case AC_OK:
			return 1;
		default:
			cmesg("Unrecognized return from configure\n", 0, 0);
			return 0;
	}
}
